[Enhancement] Improve readability of query profile

zhan7236 opened the pull request:
## Why I'm doing:

The current query profile output includes many zero-value counters and displays min/max statistics as separate counters, which makes the profile difficult to read and analyze. Users have requested improvements to simplify and restructure the query profile to enhance readability.

## What I'm doing:

This patch improves the readability of query profile output by:

1. **Omitting zero-value counters by default**
   - Counters with value=0 are now hidden to reduce noise
   - Use display_threshold < 0 to force display even if zero

2. **Displaying min/max statistics inline**
   - Min/max values are shown in format: `value [min, max]`
   - Separate `__MIN_OF_` and `__MAX_OF_` counters are filtered from output
   - Min/max is only shown when values differ from the main counter

3. **Adding `shouldDisplay()` method to Counter class**
   - Provides consistent filtering logic across BE and FE
   - Respects display_threshold settings

Fixes #60952

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds profile format versioning (v1 legacy, v2 compact with inline min/max) and counter display-threshold filtering, plus a session variable to control format and comprehensive tests.
> 
> - **Profile formatting (RuntimeProfile)**:
>   - Add `prettyPrint(..., formatVersion)` and `DefaultProfileFormatter` support for versioned output: `1` (legacy) vs `2` (compact, inlines min/max and hides `__MIN_OF_`/`__MAX_OF_`).
>   - Introduce `printCounterWithMinMax` and update child-counter printing to inline min/max in compact mode.
> - **Counters**:
>   - Add `Counter.shouldDisplay()` honoring `strategy.display_threshold` (<0/0 always show; >0 show if `value > threshold`).
> - **Session variables**:
>   - New `profile_format_version` with getter/setter to control output format.
> - **Tests**:
>   - Add UTs for display-threshold logic, min/max inline printing, pretty print format versions, and `getMaxCounter` behavior.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 43c2ea0750fff454c3ee120df5f5011d6dfb43ef. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

zhan7236 committed:
[Enhancement] Improve readability of query profile

This patch improves the readability of query profile output by:

1. Omitting zero-value counters by default
   - Counters with value=0 are now hidden to reduce noise
   - Use display_threshold < 0 to force display even if zero

2. Displaying min/max statistics inline
   - Min/max values are shown in format: value [min, max]
   - Separate __MIN_OF_ and __MAX_OF_ counters are filtered from output
   - Min/max is only shown when values differ from the main counter

3. Adding shouldDisplay() method to Counter class
   - Provides consistent filtering logic across BE and FE
   - Respects display_threshold settings

Fixes #60952

Signed-off-by: zhan7236 <76658920+zhan7236@users.noreply.github.com>

alvin-celerdata commented:
@cursor review

zhan7236 committed:
[Enhancement] Fix clang-format and add unit test for should_display

- Fix trailing whitespace in runtime_profile.h for clang-format compliance
- Add BE unit test for Counter::should_display() to improve coverage:
  - threshold < 0: always display (force show even if zero)
  - threshold == 0: display only if value != 0
  - threshold > 0: display only if value > threshold

Signed-off-by: zhan7236 <76658920+zhan7236@users.noreply.github.com>

zhan7236 committed:
[Enhancement] Fix CI failures and maintain backward compatibility

- Fix NullPointerException in Counter.shouldDisplay() when strategy is null
- Change default behavior to always display counters (threshold <= 0)
  to maintain backward compatibility with existing visualization tools
- Only hide counters when display_threshold > 0 is explicitly set
- Update test cases to reflect the new compatible behavior
- Apply clang-format fixes

This addresses the review feedback from @murphyatwork about maintaining
compatibility with existing visualization tools.

Signed-off-by: zhan7236 <76658920+zhan7236@users.noreply.github.com>

zhan7236 commented:
> `Displaying min/max statistics inline`
> 
> That's a good idea, but it might compromise the compatibility of existing visualization tools. I would suggest introducing a configuration option, such as `profile_minmax_format_version`, to allow flexibility and support changes in behavior.

@murphyatwork Thanks for the review! Good point on the compatibility issue.

I've pushed updates to fix the CI (NPE and test assertions) and tweaked the logic:

1. **Backward Compatibility**: Reverted the default behavior. Counters are now **always displayed** if `display_threshold <= 0` (the default). They are only hidden when a threshold is explicitly set. This should keep existing tools happy.
2. **Inline Display**: Min/max values are still shown inline (`value [min, max]`), but I'm filtering out the separate `__MIN_OF_` / `__MAX_OF_` counters to reduce noise.

Regarding `profile_minmax_format_version`: I feel the current `display_threshold` mechanism offers enough flexibility without adding another config variable. If you feel strongly about having a dedicated session variable, I'm happy to add it.

PTAL, thanks!

murphyatwork commented on fe/fe-core/src/main/java/com/starrocks/common/util/RuntimeProfile.java:
```diff
@@ -1001,27 +1028,26 @@
+                // Print counter with inline min/max statistics
                 builder.append(prefix).append("   - ").append(childName).append(": ")
-                        .append(printCounter(childCounter.getValue(), childCounter.getType())).append("\n");
+                        .append(printCounterWithMinMax(childCounter)).append("\n");
```
@zhan7236 you may need to resolve this

zhan7236 committed:
Address reviewer feedback: revert BE code change and restore JSON formatter compatibility

- Revert BE should_display() to original one-liner (no behavior change, no code change needed)
- Restore JsonProfileFormatter to original implementation for tool compatibility
- Keep text-only improvements in DefaultProfileFormatter

Signed-off-by: zhan7236 <76658920+zhan7236@users.noreply.github.com>

zhan7236 commented on be/src/util/runtime_profile.h:
```diff
@@ -171,9 +171,21 @@ class RuntimeProfile {
         bool should_display() const {
             int64_t threshold = _strategy.display_threshold;
-            return threshold == 0 || value() > threshold;
+            // threshold <= 0: always display (default, for compatibility)
```
Done. Reverted BE changes and restored JsonProfileFormatter for compatibility. Thanks

murphyatwork commented:
@zhan7236 i'm wondering if the `ANALYZE PROFILE` function still work? it may relies on the `MIN/MAX` metrics. if you remove the existing `MIN/MAX` metrics, these kind of tools might also need to change

zhan7236 committed:
feat: introduce profile_format_version for compact min/max display

Add session variable  to control text profile output:
- 1 (default): Legacy format with separate min/max lines.
- 2: Compact format with inline  display.

This ensures backward compatibility by default and keeps JSON format unchanged for tools.

Fixes #60952

Signed-off-by: zhan7236 <76658920+zhan7236@users.noreply.github.com>

alvin-celerdata commented:
@cursor review

zhan7236 committed:
Fix SonarCloud Optional.get() issue

Signed-off-by: zhan7236 <76658920+zhan7236@users.noreply.github.com>

zhan7236 committed:
Fix test: add missing min_max_type to TCounterStrategy

Signed-off-by: zhan7236 <76658920+zhan7236@users.noreply.github.com>

zhan7236 committed:
Fix: Add missing Optional import in RuntimeProfile.java

Add the missing java.util.Optional import that was causing
compilation failures in CI (FE Sonarcloud Check and FE UT).

Signed-off-by: zhan7236 <76658920+zhan7236@users.noreply.github.com>

stdpain commented:
https://github.com/Mergifyio rebase main
